### PR TITLE
[CU-f94x9r] Add api service

### PR DESCRIPTION
**Pull request by Zummek**

### CODE REVIEW

**Comment by Zummek** [comment]
Task linked: [CU-f94x9r Create API connection service](https://app.clickup.com/t/f94x9r)

File: web-app/src/services/http.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+import { get, isNil } from 'lodash';
+import { Notify } from 'quasar';
+import internalApi from './internalApi';
+
+const api = {
+  internalApi,
+};
+
+export default api;
+export const jc = axios.create({
```
Comment (eastemployeer) [inline web-app/src/services/http.ts:11]: jc? What does it mean? Variable should have proper name

File: web-app/src/services/http.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  if (
+    'TokenManager.getAccessToken() - get from cookie' &&
+    config.url !== 'refresh'
+  ) {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+    config.headers['Authorization'] =
+      'Bearer ' + 'TokenManager.getAccessToken()';
+  }
+  return config;
+};
```
Comment (eastemployeer) [inline web-app/src/services/http.ts:29]: What this function actually does is extracting token from cookie and sending it as a Bearer token which is totally WRONG! Cookies are automatically sent alongside request and they DON'T need to (actually it is  MISTAKE)  be converted to Bearer token. It's distracting whole philosophy of using cookies with request. We want cookies so we use cookies, not Bearer tokens!!!

**Review by eastemployeer — CHANGES_REQUESTED** [review]
Whole http.ts file needs correction. I stopped CR after I saw this crime commited on cookies...

**Comment by eastemployeer** [comment]
Merge requested. @Zummek

**Review by eastemployeer — APPROVED** [review]

**Comment by Zummek** [comment]
@eastemployeer you were to make changes

File: web-app/src/pages/Login.vue [hunk 3]
```diff
@@ -69,6 +70,11 @@ export default class Login extends Vue {
           code: this.$route.query.code,
         },
       });
+      // private async mounted() {
+      //   if (this.$route.query.code) {
+      //     await api.internalApi.auth.getToken(String(this.$route.query.code));
```
Comment (eastemployeer) [inline web-app/src/pages/Login.vue:75]: Do we really need this kind of code structure? `api.internalApi.auth.getToken`. We can access `getToken` directly, without accessing `api.internalApi.auth`. Also you are making two-side imports/exports in http.ts, auth.ts, index.ts and so on. I don't really see sense of it.

**Review by eastemployeer — COMMENTED** [review]
Need to discuss structure of API service and these two-side imports/exports everywhere. @Zummek

File: web-app/src/services/API/internalAPI/apiResources/types.ts [hunk 4]
```diff
@@ -1,6 +1,11 @@
+import { AxiosResponse } from 'axios';
+
 export interface AuthLoginResponse {
   data: Location;
 }
+export interface LogoResponse extends AxiosResponse {
```
Comment (Zummek) [inline web-app/src/services/API/internalAPI/apiResources/types.ts]: this extends look useless

File: web-app/src/services/API/internalAPI/axiosInstance.ts [hunk 5]
```diff
@@ -18,9 +18,8 @@ const handleResponseErrors = (response: AxiosResponse) => {
     | number;
 
   // TODO: check cookie
```
Comment (Zummek) [inline web-app/src/services/API/internalAPI/axiosInstance.ts]: Remove TODO comment

File: web-app/src/services/API/internalAPI/axiosInstance.ts [hunk 6]
```diff
@@ -36,7 +35,8 @@ const handleResponseErrors = (response: AxiosResponse) => {
       break;
     case 401:
       if (currentlyLoggedIn) {
-        // TODO: remove here cookie
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+        response.headers['set-cookie'] = undefined;
```
Comment (Zummek) [inline web-app/src/services/API/internalAPI/axiosInstance.ts]: Here You should remove a cookie from browser
https://github.com/cmp-cc/vue-cookies#api

File: web-app/src/services/API/internalAPI/axiosInstance.ts [hunk 7]
```diff
@@ -18,9 +18,8 @@ const handleResponseErrors = (response: AxiosResponse) => {
     | number;
 
   // TODO: check cookie
-  const currentlyLoggedIn = !isNil(
-    'TokenManager.getAccessToken() - check cookie here'
-  );
+  // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+  const currentlyLoggedIn = !isNil(response.headers['set-cookie']);
```
Comment (Zummek) [inline web-app/src/services/API/internalAPI/axiosInstance.ts]: Here check cookie in browser, not from response/request
https://github.com/cmp-cc/vue-cookies#api

File: web-app/src/services/API/internalAPI/apiResources/auth.ts [hunk 8]
```diff
@@ -1,13 +1,14 @@
...
   const response = await axiosInstance.post<types.AuthLoginResponse>(
     'v1/users/login'
   );
-  return response.data;
+  return response.data.data;
 };
 
 export const getCookieToken = async (code: string) => {
+  axiosInstance.defaults.withCredentials = true;
   await axiosInstance.post('v1/users/logged', { code });
```
Comment (Zummek) [inline web-app/src/services/API/internalAPI/apiResources/auth.ts]:
```suggestion
await axiosInstance.post(
    'v1/users/logged',
    { code },
    { withCredentials: true }
  );
```

it is not correct. When You want to set `withCredentials` header only to this request you should do this like

File: web-app/src/services/API/internalAPI/apiResources/types.ts (hunk 4 above)
Comment (eastemployeer) [reply web-app/src/services/API/internalAPI/apiResources/types.ts]: No it is not useless. LogoResponse is still AxiosResponse but with logo field in data.

File: web-app/src/services/API/internalAPI/axiosInstance.ts (hunk 7 above)
Comment (eastemployeer) [reply web-app/src/services/API/internalAPI/axiosInstance.ts]: Yeah that seemed strange to me as well but you prepared this strange "schema" and I didn't really know what it was about.

File: web-app/src/services/API/internalAPI/apiResources/auth.ts (hunk 8 above)
Comment (eastemployeer) [reply web-app/src/services/API/internalAPI/apiResources/auth.ts]: Okey

File: web-app/src/services/API/internalAPI/apiResources/types.ts (hunk 4 above)
Comment (Zummek) [reply web-app/src/services/API/internalAPI/apiResources/types.ts]: Look on `AuthLoginResponse` this don't extend `AxiosResponse` and working :P 
When You pass return pass to `.get<type>(...)` its automatically pass type to return data

`extends AxiosResponse` is useless

File: web-app/src/store/user/mutations.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import { MutationTree } from 'vuex';
+import { UserStateInterface } from './state';
+
+const mutation: MutationTree<UserStateInterface> = {
```
Comment (Zummek) [inline web-app/src/store/user/mutations.ts:4]: Add `logOut()` mutation for logout to restore default values

File: web-app/src/store/user/getters.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+  },
+  tenantId(state: UserStateInterface) {
+    return state.tenantId;
+  },
+  name(state: UserStateInterface) {
+    return state.name;
+  },
+  username(state: UserStateInterface) {
+    return state.username;
+  },
```
Comment (Zummek) [inline web-app/src/store/user/getters.ts]: getters are for custom data. e.q. `fullName()`
raw properties are available from state e.q. `this.$store.state.currentUser.id`

File: web-app/src/store/index.ts [hunk 11]
```diff
@@ -1,27 +1,21 @@
...
+  user: UserStateInterface;
 }
 
 export default store(function({ Vue }) {
   Vue.use(Vuex);
 
   const Store = new Vuex.Store<StateInterface>({
     modules: {
-      // example
+      user: userModule,
```
Comment (Zummek) [inline web-app/src/store/index.ts]: Maybe change `user` to `currentUser` for better eloquence of the object

File: web-app/src/store/user/getters.ts (hunk 10 above)
Comment (eastemployeer) [reply web-app/src/store/user/getters.ts]: Getting raw properties is not a good practice. We should avoid it.

File: web-app/src/store/user/mutations.ts (hunk 9 above)
Comment (eastemployeer) [reply web-app/src/store/user/mutations.ts:4]: Added

**Comment by eastemployeer** [comment]
@Zummek  Merge requested

**Pull request merged**